File: backend/app/tools.py
"""Server-side tool endpoints.

The app routes every server tool the Realtime model calls to POST /tools/{name} with the
tool arguments as the JSON body, and returns the JSON result back into the conversation.

`get_profile` returns only WHICH fields are on file (no values — EU data residency); the actual
values are served by the internal `fill_values` endpoint, which the app calls on-device to inject
into forms and which is deliberately NOT advertised to the model. `search_eu_info` runs RAG over
the curated EU corpus; `web_search` hits Tavily constrained to europa.eu (Phase 4).
"""

import logging
from typing import Any

from fastapi import APIRouter, Body, Depends

from .auth import User, get_current_user
from .config import get_settings
from .preferences import save_preference
from .profile import get_active_profile, profile_manifest
from .rag import search as rag_search
from .websearch import web_search as tavily_search

log = logging.getLogger("youthbuddy.tools")

router = APIRouter(prefix="/tools", tags=["tools"])


@router.post("/get_profile")
def get_profile(user: User = Depends(get_current_user)) -> dict[str, Any]:
    """Tell the model WHICH of the user's details are on file — names only, never values.

    EU data residency: the user's personal data (from their uploaded ID) stays in Supabase and
    is inserted into forms on-device by `fill_form`. The assistant must never receive the values,
    so this returns just a field manifest. To fill a form the model calls `fill_form` (no args).
    """
    manifest = profile_manifest(get_active_profile(user.id))
    return {
        **manifest,
        "note": (
            "Field names only — the values are stored securely in the EU and inserted straight "
            "into forms; they are never shown to you. Call fill_form to use them."
        ),
    }


@router.post("/fill_values")
def fill_values(user: User = Depends(get_current_user)) -> dict[str, Any]:
    """INTERNAL — the saved field VALUES for on-device form injection.

    This is NOT one of the model's tools (it is absent from TOOL_DEFS), so the LLM cannot call
    it and never sees its output. The app's `fill_form` executor calls it directly to read the
    user's values and write them into the open form's DOM. Auth-scoped to the signed-in user.
    """
    return {"profile": get_active_profile(user.id)}


@router.post("/save_preference")
def save_preference_endpoint(
    args: dict[str, Any] = Body(default={}),
    user: User = Depends(get_current_user),
) -> dict[str, Any]:
    """Remember a light, non-sensitive preference the user volunteered (see app/preferences.py).

    Unlike the ID profile, these values are non-sensitive and are shown back to the model so it
    can personalise. The model is instructed (persona) never to put sensitive data here.
    """
    prefs = save_preference(user.id, args.get("key", ""), args.get("value", ""))
    return {"ok": True, "preferences": prefs}


@router.post("/search_eu_info")
async def search_eu_info(
    args: dict[str, Any] = Body(default={}),
    user: User = Depends(get_current_user),
) -> dict[str, Any]:
    """RAG over the EU knowledge base (seed corpus + crawled portals): chunks + source URLs."""
    query = args.get("query", "")
    k = int(args.get("k", 5))
    try:
        results = await rag_search(get_settings(), query, k=max(1, min(k, 8)))
    except RuntimeError as e:
        log.error("search_eu_info failed: %s", e)
        return {"query": query, "results": [], "note": str(e)}
    return {"query": query, "results": results}


@router.post("/web_search")
async def web_search(
    args: dict[str, Any] = Body(default={}),
    user: User = Depends(get_current_user),
) -> dict[str, Any]:
    """Live web search constrained to europa.eu via Tavily."""
    query = args.get("query", "")
    return await tavily_search(get_settings(), query)

File: mobile/src/tools.ts
// Tool definitions + executor/router for the Realtime model.
//
// Definitions are sent to the model via session.update on connect. When the model calls a
// tool, executeTool() routes it:
//   • client tools (open_form, fill_form, read_page) -> WebView (via ctx)
//   • server tools (get_profile, search_eu_info, web_search) -> POST {backendUrl}/tools/{name}
// and returns a string that the app feeds back as the function_call_output.
//
// EU data residency: fill_form reads the saved values from the INTERNAL /tools/fill_values
// endpoint (not advertised to the model) and injects them straight into the WebView. Those
// values are never returned to the model — get_profile only ever exposes field NAMES.

import { EligibilityProfile, buildInjectedJavaScript, buildReadPageJavaScript } from './injection';
import { COUNTRY_OPTIONS } from './countryOptions';

// Map a country the model said (a code like "RO" OR a name like "Romania"/"france") to the
// exact Drupal <option> value. The model usually says the name; the form needs the code.
const NAME_TO_CODE: Record<string, string> = Object.entries(COUNTRY_OPTIONS).reduce(
  (acc, [code, label]) => {
    acc[label.toLowerCase()] = code;
    return acc;
  },
  {} as Record<string, string>
);

function resolveCountryCode(input: string): string {
  if (!input) return input;
  if (COUNTRY_OPTIONS[input]) return input; // already a valid code, e.g. "RO"
  const byName = NAME_TO_CODE[String(input).toLowerCase().trim()];
  return byName || input; // fall back to raw (fill will report a clear error if invalid)
}

// ---- Tool schemas advertised to the Realtime model ----
export const TOOL_DEFS = [
  {
    type: 'function',
    name: 'open_form',
    description:
      'Open an official EU web form in the in-app browser so it can be filled. Use the European Solidarity Corps eligibility form when the user wants to sign up.',
    parameters: {
      type: 'object',
      properties: {
        url: { type: 'string', description: 'The form URL to open.' },
      },
      required: ['url'],
    },
  },
  {
    type: 'function',
    name: 'fill_form',
    description:
      "Fill the currently open form from the user's saved details. Call it with no arguments — the app reads the saved values on-device and inserts them; you do NOT see the values. Only pass an argument to override a field the user just told you out loud.",
    parameters: {
      type: 'object',
      properties: {
        country: { type: 'string', description: 'Optional override, e.g. "RO" for Romania.' },
        birthdate: { type: 'string', description: 'Optional override as yyyy-mm-dd.' },
      },
      required: [],
    },
  },
  {
    type: 'function',
    name: 'read_page',
    description:
      'Inspect the open form: which fields exist, current values, and valid country option codes.',
    parameters: { type: 'object', properties: {} },
  },
  {
    type: 'function',
    name: 'get_profile',
    description:
      "Check WHICH of the user's details are on file (returns field names only, never the values — they stay private in the EU). To fill a form, call fill_form; it uses the saved values directly without showing them to you.",
    parameters: { type: 'object', properties: {} },
  },
  {
    type: 'function',
    name: 'search_eu_info',
    description: 'Search indexed official EU sources for youth programmes, rights and funding.',
    parameters: {
      type: 'object',
      properties: { query: { type: 'string' } },
      required: ['query'],
    },
  },
  {
    type: 'function',
    name: 'web_search',
    description: 'Search the live web (official europa.eu domains) for up-to-date EU information.',
    parameters: {
      type: 'object',
      properties: { query: { type: 'string' } },
      required: ['query'],
    },
  },
  {
    type: 'function',
    name: 'save_preference',
    description:
      "Remember a light, NON-sensitive preference the user mentions (e.g. key='climate', value='warm'). Never store sensitive personal data (ID, address) this way.",
    parameters: {
      type: 'object',
      properties: {
        key: { type: 'string', description: "Short preference name, e.g. 'climate'." },
        value: { type: 'string', description: "Short value, e.g. 'warm'." },
      },
      required: ['key', 'value'],
    },
  },
] as const;

export const CLIENT_TOOLS = new Set(['open_form', 'fill_form', 'read_page']);

// The default form the buddy signs people up on.
export const ESC_ELIGIBILITY_URL = 'https://youth.europa.eu/solidarity/register/check_en';

/** Everything executeTool needs from the app, so the router itself stays testable. */
export interface ToolContext {
  backendUrl: string;
  newRequestId: () => string;
  /** Navigate the WebView; resolves once navigation is requested. */
  openForm: (url: string) => Promise<void>;
  /** Inject JS and resolve with the postMessage payload matching requestId. */
  injectAndWait: (js: string, requestId: string) => Promise<any>;
  /** POST {backendUrl}/tools/{name} with args; resolve parsed JSON. */
  callServerTool: (name: string, args: any) => Promise<any>;
}

/** Execute one tool call and return the string output to hand back to the model. */
export async function executeTool(name: string, args: any, ctx: ToolContext): Promise<string> {
  args = args || {};
  switch (name) {
    case 'open_form': {
      const url = args.url || ESC_ELIGIBILITY_URL;
      await ctx.openForm(url);
      return JSON.stringify({ ok: true, opened: url });
    }
    case 'read_page': {
      const id = ctx.newRequestId();
      const msg = await ctx.injectAndWait(buildReadPageJavaScript(id), id);
      return JSON.stringify(msg.result ?? { error: msg.error ?? 'no result' });
    }
    case 'fill_form': {
      // EU data residency: read the saved values from the INTERNAL fill_values endpoint (not a
      // model tool — the LLM never sees this) and inject them directly. The model may pass an
      // override for a field the user said out loud; otherwise everything comes from the vault.
      let country: string = args.country;
      let birthdate: string = args.birthdate;
      if (!country || !birthdate) {
        const p = await ctx.callServerTool('fill_values', {});
        country = country || p.profile.country;
        birthdate = birthdate || p.profile.birthdate;
      }
      const profile: EligibilityProfile = { country: resolveCountryCode(country), birthdate };
      const id = ctx.newRequestId();
      const msg = await ctx.injectAndWait(buildInjectedJavaScript(profile, id), id);
      return JSON.stringify(msg.result ?? { error: msg.error ?? 'no result' });
    }
    case 'get_profile':
    case 'search_eu_info':
    case 'web_search':
    case 'save_preference': {
      const result = await ctx.callServerTool(name, args);
      return JSON.stringify(result);
    }
    default:
      return JSON.stringify({ error: `unknown tool ${name}` });
  }
}
